### DISCUSSION TITLE
Is Context propagation support over web socket?

### DISCUSSION THREAD

**Discussion by ychaydhari** [open-telemetry/community#1215] [Community Discussions]
Hello,
We have two services running on the machine. Both services communicate with each other through web socket. Can we propagate trace context through the web socket? or do you have any examples for reference?

**Comment by SergeyKanzhelev** [comment]
@ychaydhari do you have a specific language and framework in mind? Framework may help with metadata like trace context encoding with the message. Also the semantic of something being a single trace and other things being spans or events on the span may not be clear without some framework around websocket communication.

**Comment by tigrannajaryan** [comment]
AFAIK, there is no standardized way to attach arbitrary metadata to individual WebSocket messages (like e.g. HTTP does), so I don't see how Otel can have a generic support for context propagation. It will need to be implemented per WebSocket-based protocol and the context propagation functionality needs to be aware of how individual messages in that particular protocol are encoded so that the context can be attached to the payload that goes into the message. In other words if you have an in-house WebSocket-based protocol you have to extend your protocol to allow encoding the context in your protocol's message payload.

If the _entire_ WebSocket connection reflects one trace context then the context can be propagated via regular HTTP headers (using w3c context headers) which are passed when the WebSocket connection is established. However, this is likely not very useful except for some very narrow use cases with short-lived WebSocket connections.

**Comment by psk001** [comment]
HI @tigrannajaryan , I have been using opentelemetry for tracing in our distributed services which have nodeJS backend, use opentelemetry/sdk-trace-base for exporters and collector hosted on k8s. It works well for intra-service context propagation. But when it comes to inter -services context propagation, i did not find any suitable documentation. I tried passing it through headers as mentioned in [W3C docs](https://www.w3.org/TR/trace-context/) but i was not able to pass the tracecontext. Are there any references like demo service which I can reference and take help to set up distributed tracing? Thanks and cheer

**Reply by tigrannajaryan** [reply]
The demo here https://github.com/open-telemetry/opentelemetry-demo has multiple services, so maybe what you are looking for.
